src/wp-includes/html-api/class-wp-html-processor.php

```diff
@@ -287,12 +287,14 @@ class WP_HTML_Processor extends WP_HTML_Tag_Processor {
 	 * @since 6.4.0
 	 * @since 6.6.0 Returns `static` instead of `self` so it can create subclass instances.
 	 *
-	 * @param string $html     Input HTML fragment to process.
-	 * @param string $context  Context element for the fragment, must be default of `<body>`.
-	 * @param string $encoding Text encoding of the document; must be default of 'UTF-8'.
+	 * @param string $html          Input HTML fragment to process.
+	 * @param string $context       Context element for the fragment, must be default of `<body>`.
+	 * @param string $encoding      Optional. Text encoding of the document; must be default of 'UTF-8'.
+	 * @param string $document_mode Optional. Set document compatibility mode (quirks). Should be
+	 *                              'quirks-mode' or 'no-quirks-mode'. Default: 'no-quirks-mode'.
 	 * @return static|null The created processor if successful, otherwise null.
 	 */
-	public static function create_fragment( $html, $context = '<body>', $encoding = 'UTF-8' ) {
+	public static function create_fragment( $html, $context = '<body>', $encoding = 'UTF-8', $document_mode = WP_HTML_Processor_State::NO_QUIRKS_MODE ) {
```

Comment:
A fragment parser would typically inherit the context element's document's compatibility mode.

I suspect we'll need to change how the context element is passed, but I don't think it will include information about it's document mode and this argument will remain helpful.

See #7141.

Reply:
do you think this is worth supporting right now? do we have enough warrant to include it? in what cases would we want to create a document fragment in quirks mode, and in those cases, how would we know?

I can definitely see value in having this, but also I wonder if this inclusion will help people understand better what they need to be doing or add confusion.

what do you think the consequences would be of simply not supporting a quirks-mode fragment parser? or at least of not having it in the primary function signature for creating the class?

Reply:
This was helpful and important for testing while working on these changes.

> In what cases would we want to create a document fragment in quirks mode, and in those cases, how would we know?

The fragment should use the same mode as the document for the context element. This should become clear when we work on `set_{inner,outer}_html` where we'll be creating fragment parsers that use the parent parser's document mode.

Another option would be to adjust the full parser to handle doctype declarations and set the document_mode correctly according to the full HTML document. Then we could certainly omit these changes.

I'm also reluctant to eagerly add more to this method signature.

Reply:
> what do you think the consequences would be of simply not supporting a quirks-mode fragment parser? 

If the full parser supports quirks mode and we have `set_{inner,outer}_html` methods, I think the fragment parser must support quirks mode.

> or at least of not having it in the primary function signature for creating the class?

The mode should be based on the context element's document's mode. There's no way to provide that information to the fragment parser right now. The context element is currently passed to the fragment parser as an HTML-like string (only `<body>` is allowed right now) which [seems insufficient to pass all the information](https://github.com/WordPress/wordpress-develop/pull/7141#issuecomment-2283867964) the fragment parser requires.

There are other ways to handle this, for example an instance method could create a fragment from a node and set quirks mode appropriately as well as handle things like reading attributes, namespace, etc. This is all best discussed in #7141.

I am going to explore a change to handle quirks mode in the full parser based on the doctype declaration, that would be sufficient and we could remove this change.

Reply:
I could see some value in having a method like `->set_quirks_mode()`

I still suspect that it'll be basically universal that we don't work in the full parser mode within WordPress. almost no HTML-processing code has access to the full document, and so that's what I meant when assuming _no-quirks mode_. We just don't know even when doing `inner_html` operations if the parent document is in quirks mode or not. "We can only assume UTF-8, no-quirks, `<body>` context."

Reply:
There may be value in allowing the compat mode to be changed on a processor. I'd like to leave those changes for consideration in their own PR. There was no way to change the compat mode before and I don't think we need to address that in this PR.